File: modules/programming/app/model/ProgEvaluationResult.java
package model;

import java.util.List;

import model.execution.AExecutionResult;
import model.execution.SyntaxError;
import model.exercise.Success;
import model.result.EvaluationResult;
import model.testdata.ITestData;

public class ProgEvaluationResult extends EvaluationResult {

  private AExecutionResult executionResult;
  private ITestData testData;

  public ProgEvaluationResult(AExecutionResult theExecutionResult, ITestData theTestData) {
    super(analyze(theExecutionResult, theTestData));
    executionResult = theExecutionResult;
    testData = theTestData;
  }

  private static Success analyze(AExecutionResult executionResult, ITestData testData) {
    if(executionResult instanceof SyntaxError)
      return Success.NONE;

    return validateResult(executionResult.getResult(), testData.output) ? Success.COMPLETE : Success.NONE;
  }

  protected static <T> boolean validateResult(T gottenResult, T awaitedResult) {
    // FIXME: validation of result is dependent on language! Example: numbers in
    return gottenResult.equals(awaitedResult);
  }

  public String getAwaitedResult() {
    return testData.output;
  }

  public AExecutionResult getExecutionResult() {
    return executionResult;
  }

  public int getId() {
    return testData.getId();
  }

  public List<String> getInput() {
    return testData.getInput();
  }

  public String getRealResult() {
    return executionResult.getResult();
  }

  // public String getAsHtml() {
  // FIXME: Auto-generated method stub
  // boolean successful = success == Success.COMPLETE;
  //
  // StringBuilder builder = new StringBuilder();
  // builder.append("<div class=\"col-md-6\">\n");
  // builder.append(" <div class=\"alert alert-" + (successful ? "success" :
  // "danger") + "\">\n");
  // builder.append(" <p>Test von <code>" + executionResult.getEvaluated() +
  // "</code> war "
  // + (successful ? "" : "nicht ") + "erfolgreich.<p>\n");
  // builder.append(" <p>Erwartet: \"" + testData.output + "\"</p>\n");
  // builder.append(" <p>Ergebnis: \"" + executionResult.getResult() +
  // "\"</p>\n");
  // builder.append(" " + DIV_END + "\n" + DIV_END + "");
  // return builder.toString();
  // }

  public String getTitleForValidation() {
    if(executionResult instanceof SyntaxError)
      return ((SyntaxError) executionResult).getErrorMessage();

    if(isSuccessful())
      return "Dieser Test war erfolgreich.";

    return "Dieser Test war nicht erfolgreich.\nErwartetes Ergebnisses: " + testData.output + "\nReales Ergebnis: "
        + executionResult.getResult();
  }

  @Override
  public String toString() {
    return testData.toString() + ", " + executionResult.getResult();
  }

}

File: modules/programming/app/model/ProgEvaluationResult.scala
package model

import model.ProgEvaluationResult._
import model.result.{EvaluationResult, SuccessType}
import model.testdata.ITestData

abstract sealed class AExecutionResult(val evaluated: String, val result: String, val consoleOutput: String)

case class ExecutionResult(e: String, r: String, c: String) extends AExecutionResult(e, r, c)

case class SyntaxError(e: String, error: String, c: String) extends AExecutionResult(e, error, c)

object ProgEvaluationResult {
  private def analyze(executionResult: AExecutionResult, testData: ITestData): SuccessType = executionResult match {
    case SyntaxError(_, _, _) => SuccessType.NONE
    case ExecutionResult(executed, result, console) => if (validateResult(result, testData.output)) SuccessType.COMPLETE else SuccessType.NONE
  }

  private def validateResult[T](gottenResult: T, awaitedResult: T): Boolean = {
    // FIXME: validation of result is dependent on language! Example: numbers in
    gottenResult == awaitedResult
  }
}

case class ProgEvaluationResult(executionResult: AExecutionResult, testData: ITestData) extends EvaluationResult(analyze(executionResult, testData)) {

  def awaitedResult: String = testData.output

  def id: Int = testData.getId

  def input: java.util.List[String] = testData.getInput

  def realResult: String = executionResult.result

  def getTitleForValidation: String = executionResult match {
    case SyntaxError(_, error, _) => error
    case ExecutionResult(_, _, _) => if (isSuccessful)
      "Dieser Test war erfolgreich."
    else
      s"""Dieser Test war nicht erfolgreich.
         |  Erwartetes Ergebnisses: ${testData.output}
         |  Reales Ergebnis: " + executionResult.result""".stripMargin
  }

  override def toString: String = testData.toString + ", " + executionResult.result

}
